clang/lib/CodeGen/CGDecl.cpp

```diff
@@ -1759,20 +1759,29 @@ void CodeGenFunction::emitZeroOrPatternForAutoVarInit(QualType type,
                                                       const VarDecl &D,
                                                       Address Loc) {
   auto trivialAutoVarInit = getContext().getLangOpts().getTrivialAutoVarInit();
+  auto trivialAutoVarInitSizeBound =
+      getContext().getLangOpts().TrivialAutoVarInitSizeBound;
   CharUnits Size = getContext().getTypeSizeInChars(type);
   bool isVolatile = type.isVolatileQualified();
   if (!Size.isZero()) {
+    auto allocSize = CGM.getDataLayout().getTypeAllocSize(Loc.getElementType());
```

Comment:
After second thought 
what if we have a large array of structs with a few uninitialized fields.
Initializing the rest will cost us anyway about the same.

Would it be better to check insize of `emitStoresForZeroInit`/`emitStoresForPatternInit`, more precicely in `emitStoresForConstant` before memset if IsAutoInit?

Reply:
The problem of `emitStoresForConstant` is that it may split struct/array variables to a handful of stores, which would mess up the size filtering (e.g., difficult to debug). What do you think?

> what if we have a large array of structs with a few uninitialized fields.
Initializing the rest will cost us anyway about the same.

(Correct me if I'm wrong) While emitting the auto-init, clang has no idea how each filed/element is initialized. It's difficult to distinguish these two cases in `CGDecl.cpp`:
- A large array of structs with all fields uninitialized.
- A large array of structs with a few field uninitialized.

Reply:
Ignoring a few less common cases (C++ classes with non-trivial constructors, unions, self-referencing initializers), variable initialization is generally is an all-or-nothing proposition. If you write something like `struct X { int x,y,z; }; struct X x[100] = {0};`, that initializes every field of every element.  What examples are you worried about?

Reply:
`struct S {
  long a;
  long b;
  char c;
}; S var;` The `var` size is 24 bytes. `emitStoresForConstant` splits the constant store and emits several stores instead. Assume we set the max-size flag to 16. If filter the size in `emitStoresForConstant`, this var will be auto-initialized (each split store is less than 16) even though its size > 16. Would it be confusing?

Reply:
I'd expect that it's kind of `bail out` measure which should not be consider as a contract by a end-user. So confusing should not be a concern.

Also how 
```
struct S {
  long a;
  long b;
  char c;
} var;
```

is better  then:
```
long a;
long b;
char c;
```

My impression was that your concern is buffers of hundreds of bytes, then maybe better to target just them?
If it 's `24 vs 16` maybe it should not use auto init at all?

Reply:
Thank you both for clarifying! Moved the filtering to `emitStoresForConstant` before memset.
Please take another look :-D